builtin/logical/database/versioning.go

```diff
@@ -0,0 +1,292 @@
+package database
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/helper/random"
+	"github.com/hashicorp/vault/sdk/database/dbplugin"
+	"github.com/hashicorp/vault/sdk/database/newdbplugin"
+	"github.com/hashicorp/vault/sdk/helper/pluginutil"
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+type databaseVersionWrapper struct {
+	database       newdbplugin.Database
+	legacyDatabase dbplugin.Database
```

Comment:
Even though there's the intent to leave the v4 nomenclature out out of code references, I think that terms like legacy, old, new, or similar have its own set of problems too. If we were to introduce the next iteration of database plugins (which is unlikely but not impossible), this terminology might make things confusing as it would be come the "legacy" database, and the `legacyDatabase` is really a `oldLegacyDatabase`, which would be just as ambiguous if not more so.

Reply:
Unfortunately I don't have a good way of differentiating between versions other than referring to the gRPC version number. I've switched this over to using that nomenclature within the version wrapping struct since it has more forwards compatibility.